### PR TITLE
feat: more upgrades

### PR DESCRIPTION
**Pull request by Ugzuzg**
- feat: upgrade to latest mongoose
- feat: use ajv instead of openapi-types

### CODE REVIEW

File: packages/core/src/route/types.ts [hunk 1]
```diff
@@ -85,19 +89,22 @@ type IDecoratorMetadata<T> = {
  * @param typeFactory - Function that returns a type (useful for lazily evaluated types)
  * @param required - Whether the property is required or not
  */
-export interface IPropDecoratorOptions extends Omit<IJsonSchema, 'type' | 'required'> {
-	type?: TypeValue;
+export type IPropDecoratorOptions<T> = Omit<JSONSchemaType<T, true>, 'required' | 'type'> & {
+	type?: TypeValue | JSONSchemaType<T>['type'];
```
Comment (pierissimo) [inline packages/core/src/route/types.ts:93]: out of curiosity, what's the benefit of passing the generic type to `JSONSchemaType`?

Comment (Ugzuzg) [reply packages/core/src/route/types.ts]: Example here: https://github.com/HPInc/davinci/pull/51/files#diff-5db34c11902eb9d0183002af2ec1ab37668bd589c46e606d6d7281952c4cf40cR25
If I specified the generic type, it won't allow me to add non-related options (e.g. `oneOf: [ { type: 'object' } ]` won't be accepted).

File: packages/core/src/route/openapi/createSchemaDefinition.ts [hunk 2]
```diff
@@ -116,7 +116,7 @@ export const getSchemaDefinition = (theClass: Function, definitions = {}): ISwag
 
 			return hasDefinitionDecoration
 				? {
-					$ref: `#/definitions/${title}`
+						$ref: `#/definitions/${title}`
```
Comment (pierissimo) [inline packages/core/src/route/openapi/createSchemaDefinition.ts:119]: indentation here looks weird

File: packages/core/src/route/types.ts [hunk 3]
```diff
@@ -85,19 +89,23 @@ type IDecoratorMetadata<T> = {
...
-export type IPropDecoratorMetadata = IDecoratorMetadata<IPropDecoratorOptions>;
+};
+
+export type IPropDecoratorOptionsFactory<T> = IDecoratorOptionsFactory<IPropDecoratorOptions<T>>;
+export type IPropDecoratorMetadata<T> = IDecoratorMetadata<IPropDecoratorOptions<T>>;
 
-export type IDefinitionDecoratorOptions = { title: any; } & Omit<IJsonSchema, 'type' | 'required'>;
+export type IDefinitionDecoratorOptions<T = Known> = { title: any } & Pick<
+	Partial<JSONSchemaType<T, true>>,
+	'dependencies' | 'oneOf' | 'allOf' | 'anyOf' | 'not'
```
Comment (pierissimo) [inline packages/core/src/route/types.ts:107]: why only those 5 openapi props?

File: packages/mongoose/src/hooks.ts [hunk 4]
```diff
@@ -97,27 +95,37 @@ const createHandlerArgs = (
 		isReadHook: boolean;
 		isWriteHook: boolean;
 		isDeleteHook: boolean;
-		thisObj: Document | Query<QueryResultType, Document>;
+		thisObj: Document | Query<ResultType, ResultType & Document>;
 		result?: any;
 		rest?: unknown[];
-		context?: unknown;
+		context?: Context;
```
Comment (pierissimo) [inline packages/mongoose/src/hooks.ts:101]: good spot

**Review by pierissimo — CHANGES_REQUESTED** [review]

File: packages/core/src/route/openapi/createSchemaDefinition.ts (hunk 2 above)
Comment (Ugzuzg) [reply packages/core/src/route/openapi/createSchemaDefinition.ts]: https://github.com/HPInc/davinci/pull/51/commits/e87a70e405a3478924b0149a9cc0df3546b2ccbb

File: packages/core/src/route/types.ts (hunk 3 above)
Comment (Ugzuzg) [reply packages/core/src/route/types.ts]: https://github.com/HPInc/davinci/pull/51/commits/499b8e07c7b0b75eee74b883f3d5c4d5c1ed3ee8

**Review by pierissimo — APPROVED** [review]

**Pull request merged**